rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java

```diff
@@ -0,0 +1,145 @@
+import io.seata.rm.datasource.undo.KeywordChecker;
+import io.seata.rm.datasource.undo.KeywordCheckerFactory;
+import io.seata.sqlparser.SQLRecognizer;
+import io.seata.sqlparser.SQLUpdateRecognizer;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.StringJoiner;
+
+/**
+ * The type MultiSql executor.
+ *
+ * @param <T> the type parameter
+ * @param <S> the type parameter
+ * @author wangwei-ying
+ */
+public class MultiUpdateExecutor<T, S extends Statement> extends AbstractDMLBaseExecutor<T, S> {
+    public MultiUpdateExecutor(StatementProxy<S> statementProxy, StatementCallback<T, S> statementCallback, List<SQLRecognizer> sqlRecognizers) {
+        super(statementProxy, statementCallback, sqlRecognizers);
+    }
+
+    @Override
+    protected TableRecords beforeImage() throws SQLException {
+        if (sqlRecognizers.size() == 1) {
+            UpdateExecutor executor = new UpdateExecutor<>(statementProxy, statementCallback, sqlRecognizers.get(0));
+            return executor.beforeImage();
+        }
+        final KeywordChecker keywordChecker = KeywordCheckerFactory.getKeywordChecker(getDbType());
+        final TableMeta tmeta = getTableMeta(sqlRecognizers.get(0).getTableName());
+
+        final ArrayList<List<Object>> paramAppenderList = new ArrayList<>();
+        Set<String> updateColumnsSet = new HashSet<>();
+        final StringBuilder whereCondition = new StringBuilder();
+        for (SQLRecognizer recognizer : sqlRecognizers) {
+            sqlRecognizer = recognizer;
+            SQLUpdateRecognizer sqlUpdateRecognizer = (SQLUpdateRecognizer) recognizer;
+            List<String> updateColumns = sqlUpdateRecognizer.getUpdateColumns();
+            updateColumnsSet.addAll(updateColumns);
+            String whereConditionStr = buildWhereCondition(sqlUpdateRecognizer, paramAppenderList);
+            if (whereCondition.length() > 0) {
+                whereCondition.append(" OR ");
+            }
+            if (StringUtils.isNotBlank(whereConditionStr)) {
+                whereCondition.append(whereConditionStr);
+            }
+
+        }
+        StringBuilder prefix = new StringBuilder("SELECT ");
+        if (!tmeta.containsPK(new ArrayList<>(updateColumnsSet))) {
+            prefix.append(getColumnNameInSQL(tmeta.getEscapePkName(getDbType()))).append(", ");
+        }
+        final StringBuilder suffix = new StringBuilder(" FROM ").append(getFromTableInSQL());
+        suffix.append(" WHERE ").append(whereCondition);
+        suffix.append(" FOR UPDATE");
+        final StringJoiner selectSQLAppender = new StringJoiner(", ", prefix, suffix.toString());
+        for (String column : updateColumnsSet) {
+            selectSQLAppender.add(getColumnNameInSQL(keywordChecker.checkAndReplace(column)));
```

Comment:
From the user's sql, no need to deal with keywords

Reply:
done